### PR TITLE
Domain coordinate types

### PR DESCRIPTION
**Pull request by jakebolewski**
part of #81

### CODE REVIEW

**Comment by jakebolewski** [comment]
This will close #13  minus the metadata part, as the typed geometry point type is equivalent to a "labeled" dimension.

**Comment by jakebolewski** [comment]
bors try

**Comment by jakebolewski** [comment]
bors try

**Comment by jakebolewski** [comment]
Needs a review and squash, I largely didn't touch the unstructured mesh components as they don't quite align with the previous abstractions and will need a bit more work (but largely are self contained).

**Comment by jakebolewski** [comment]
bors try

File: src/Geometry/coordinates.jl [hunk 1]
```diff
@@ -43,32 +52,75 @@ end
 
 @pointtype Cartesian12Point x1 x2
 @pointtype Cartesian123Point x1 x2 x3
+
 # TODO: deprecate these
```
Comment (simonbyrne) [inline src/Geometry/coordinates.jl:56]: can remove the TODO

File: src/Geometry/coordinates.jl [hunk 2]
```diff
@@ -43,32 +52,75 @@ end
...
-    Cartesian3DPoint{FT}(pt.x, pt.y, pt.z)
-
-components(p::AbstractPoint) = SVector(ntuple(i -> getfield(p, i), nfields(p)))
+Cartesian12Point(pt::XYPoint{FT}) where {FT} = Cartesian12Point{FT}(pt.x, pt.y)
+Cartesian123Point(pt::XYZPoint{FT}) where {FT} =
+    Cartesian123Point{FT}(pt.x, pt.y, pt.z)
+
+ncomponents(p::AbstractPoint) = nfields(p)
+components(p::AbstractPoint{FT}, i::Symbol) where {FT} = getfield(p, i)::FT
+components(p::AbstractPoint{FT}, i::Integer) where {FT} = getfield(p, i)::FT
```
Comment (simonbyrne) [inline src/Geometry/coordinates.jl:82]: Should this be singular?
```suggestion
component(p::AbstractPoint{FT}, i::Symbol) where {FT} = getfield(p, i)::FT
component(p::AbstractPoint{FT}, i::Integer) where {FT} = getfield(p, i)::FT
```

File: src/Geometry/coordinates.jl [hunk 3]
```diff
@@ -43,32 +52,75 @@ end
...
+    _coordinate_type(ptyp, Val(ax))
+coordinate_type(ptyp::Type{<:AbstractPoint}, ax::Integer) =
+    _coordinate_type(ptyp, Int(ax))
+
+coordinate(pt::AbstractPoint, ax::Int) = _coordinate(pt, Val(ax))
+coordinate(pt::AbstractPoint, ax::Integer) = _coordinate(pt, Int(ax))
+
+# TODO: we need to rationalize point operations with vectors
+Base.:(-)(p1::T, p2::T) where {T <: AbstractPoint} =
+    unionalltype(T)((components(p1) - components(p2))...)
```
Comment (simonbyrne) [inline src/Geometry/coordinates.jl:116]: What does this return?

**Review by simonbyrne — APPROVED** [review]

File: src/Geometry/coordinates.jl (hunk 3 above)
Comment (jakebolewski) [reply src/Geometry/coordinates.jl:116]: This returns a point that promotes correctly. I'll see if this is still needed as I was trying to get the local geometry to work correctly with CoordinateTypes but there needs to be a more incremental approach as there are a lot of baked in assumptions currently.

Ideally this should return a vector type.

File: src/Geometry/coordinates.jl (hunk 2 above)
Comment (jakebolewski) [reply src/Geometry/coordinates.jl:82]: I was thinking coordinates is kind of like `Base.axes` and there are not singular versions for the later, I can change it though although the similarity of the names `comoponents` `component` might be a bit error prone.

**Comment by jakebolewski** [comment]
bors try

File: src/Geometry/coordinates.jl (hunk 3 above)
Comment (simonbyrne) [reply src/Geometry/coordinates.jl]: A vector type would make sense in a box, but wouldn't e.g. on a sphere.

Where is addition/subtraction of coordinates used? Is it just for ranges?

**Comment by jakebolewski** [comment]
I added a comment it is just for range support

**Pull request merged**